codec-http/src/main/java/io/netty/handler/codec/http/DefaultHttpHeaders.java

```diff
@@ -94,26 +96,42 @@ public CharSequence convertObject(Object value) {
             return seq;
         }
 
+        private static final class ValidateValueProcessor implements ByteVisitor {
+            private int state;
+            @Override
+            public boolean process(byte value) throws Exception {
+                state = validateValueChar(state, (char) value);
+                return true;
+            }
+
+            public int state() {
+                return state;
+            }
+        }
+
         private static void validateValue(AsciiString seq) {
-            int state = 0;
-            // Start looping through each of the character
-            final int start = seq.arrayOffset();
-            final int end = start + seq.length();
-            final byte[] array = seq.array();
-            for (int index = start; index < end; index++) {
-                state = validateValueChar(seq, state, (char) (array[index] & 0xFF));
+            ValidateValueProcessor processor = new ValidateValueProcessor();
+            try {
+                seq.forEachByte(processor);
+            } catch (Throwable t) {
+                PlatformDependent.throwException(
+                        new IllegalArgumentException("an illegal header value was detected: " + seq, t));
             }
 
-            if (state != 0) {
+            if (processor.state() != 0) {
                 throw new IllegalArgumentException("a header value must not end with '\\r' or '\\n':" + seq);
             }
         }
 
         private static void validateValue(CharSequence seq) {
             int state = 0;
-            // Start looping through each of the character
-            for (int index = 0; index < seq.length(); index++) {
-                state = validateValueChar(seq, state, seq.charAt(index));
+            try {
+                // Start looping through each of the character
+                for (int index = 0; index < seq.length(); index++) {
+                    state = validateValueChar(state, seq.charAt(index));
+                }
+            } catch (IllegalArgumentException e) {
+                throw new IllegalArgumentException("an illegal header value was detected: " + seq, e);
```

Comment:
I was wrestling with this and decided to keep the state down in the `ByteProcessor` objects under the assumption that most of the time we will not have an exception.  However this is a great point that another stack trace will be generated.

We have at least 2 options:
1. Do as you suggest and add the extra state and throw the exception 1 time with the string.
2. Create our own exception class and use the [Throwable constructor](https://docs.oracle.com/javase/7/docs/api/java/lang/Throwable.html#Throwable(java.lang.String,%20java.lang.Throwable,%20boolean,%20boolean) which allows us to not generate a new stack trace.

```java
    private static final class MalformedInputException extends RuntimeException {
        private static final long serialVersionUID = 2509479631043011905L;

        public MalformedInputException(String message, Throwable cause) {
            super(message, cause, true, false);
        }
    }
```

Option 2 keeps state down in what I assume is the "majority" of use cases, at the expense of introducing our own exception type....WDYT?